keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -340,47 +352,87 @@ def next(prompt, cache, index):
 
     def generate(
         self,
-        prompt,
-        max_length,
+        inputs,
+        max_length=None,
     ):
-        """Generate text.
+        """Generate text given prompt `inputs`.
 
-        This method generates text based on given `prompt`. Generation will
-        continue until `max_length` is met, and all tokens generated after
-        `end_token` will be truncated. The sampling strategy can be set in
-        the `compile` method.
+        This method generates text based on given `prompt`. The sampling used
+        for generation can be set in the `compile` method.
 
-        Args:
-            prompt: a string, string Tensor or string RaggedTensor. The prompt
-                text for generation.
-            max_length: int. The max length of generated sequence.
-        """
-        if self.preprocessor is None:
-            raise ValueError(
-                "`self.preprocessor` is `None`, please make sure "
-                "`preprocessor` is set before calling `generate`."
-            )
+        If `inputs` are a batched `tf.data.Dataset` or `batch_size` is provided,
+        outputs will be generated "batch-by-batch" and concatenated. Otherwise,
+        all inputs will be handled in a single generation pass.
 
-        # Tokenize.
-        prompt = tf.convert_to_tensor(prompt)
-        input_is_scalar = prompt.shape.rank == 0
-        prompt = prompt[tf.newaxis] if input_is_scalar else prompt
-        prompt = self.preprocessor.tokenizer(prompt)
+        If a `preprocessor` is attached to the model, `inputs` should be
+        strings and returned sequences will be strings. Otherwise, inputs should
+        be preprocessed before calling `generate()`, and returned sequences will
+        be token ids.
 
-        # Pad ragged to dense tensors.
-        padded_shape = (None, max_length)
-        min_length = tf.reduce_min(prompt.row_lengths())
-        input_mask = tf.ones_like(prompt, tf.bool).to_tensor(shape=padded_shape)
-        prompt = prompt.to_tensor(shape=padded_shape)
+        Args:
+            inputs: a string `tf.Tensor`, a `tf.data.Dataset` of strings, a
+                python string or a list of python strings. If no `preprocessor`
+                is attached to the model, inputs should instead be a nested
+                `tf.Tensor` or `tf.data.Dataset` with keys `"token_ids"` and
+                `"padding_mask"`.
+            max_length: Optional. int. The max length of the generated sequence.
```

Comment:
But what if there's no preprocessor? In that case this arg appears not to work.

Reply:
Yeah, this arg is worth discussing. `max_length` is really conceptually a preprocessing arg and does absolutely nothing if we are passing preprocessed inputs, which should already have shape `(batch_size, max_length)`.

This feels like what you would call a denomalized argument somewhat. In that `sequence_length` on the preprocessor and `max_length` here are really setting the same thing.

We could...
 - Think about removing it, though I would do that as a follow perhaps.
 - Document that it does nothing when `preprocessor is None`.
 - Throw an error if `inputs.shape[1] != max_length`.

Thoughts?

Reply:
I think I will go with the lightweight approach of documenting for now.

Reply:
Sight edit for clarity:

```python
"""
If `preprocessor` is `None`, `inputs` should be padded to the desired maximum 
length and this argument will be ignored.
"""
```